docs/plugins/composability.md

```diff
@@ -0,0 +1,584 @@
+- Change any plain component exports, such as catalog overview cards, to be
+  component extensions.
+- Stop exporting `RouteRef`s and instead pass them to `createPlugin`.
+- Stop accepting `RouteRef`s as props or importing them from other plugins,
+  instead create an `ExternalRouteRef` as a replacement, and pass it to
+  `createPlugin.`
+- Rename any other exported symbols according to the naming pattern table below.
+
+Note that removing the existing exports and configuration is a breaking change
+in any plugin. If backwards compatibility is needed the existing code be
+deprecated while making the new additions, to then be removed at a later point.
+
+### Naming Patterns
+
+Many export naming patterns have been changed to avoid import aliases and to
+clarify intent. Refer to the following table to formulate the new name:
+
+| Description          | Existing Pattern           | New Pattern     | Examples                                       |
+| -------------------- | -------------------------- | --------------- | ---------------------------------------------- |
+| Top-level Pages      | Router                     | \*Page          | CatalogIndexPage, SettingsPage, LighthousePage |
+| Entity Tab Content   | Router                     | Entity\*Content | EntityJenkinsContent, EntityKubernetesContent  |
+| Entity Overview Card | \*Card                     | Entity\*Card    | EntitySentryCard, EntityPagerDutyCard          |
+| Entity Conditional   | isPluginApplicableToEntity | is\*Available   | isPagerDutyAvailable, isJenkinsAvailable       |
+| Plugin Instance      | plugin                     | \*Plugin        | jenkinsPlugin, catalogPlugin                   |
+
+## Porting Existing Apps
+
+The first step of porting any app is to replace the root `Routes` component with
+`FlatRoutes` from `@backstage/core`. As opposed to the `Routes` component,
+`FlatRoutes` only considers the first level of `Route` components in its
+children, and provides any additional children to the outlet of the route. It
+also removes the need to append `"/*"` to paths, as it is added automatically.
+
+```diff
+const AppRoutes = () => (
+-  <Routes>
++  <FlatRoutes>
+    ...
+-    <Route path="/docs/*" element={<DocsRouter />} />
++    <Route path="/docs" element={<DocsRouter />} />
+    ...
+-  </Routes>
++  </FlatRoutes>
+);
+```
+
+The next step should be to switch from using `EntityPageLayout` to
+`EntityLayout`, as this can also be done without waiting for plugins to be
+ported. You should also replace the top-level `Router` from the catalog plugin
+with the separate `CatalogIndexPage` and `CatalogEntityPage` extensions that
+have been added to the catalog:
+
+```diff
+-<Route
+-  path={`${catalogRouteRef.path}/*`}
+-  element={<CatalogRouter EntityPage={EntityPage} />}
+-/>
++<Route path="/catalog" element={<CatalogIndexPage />} />
++<Route
++  path="/catalog/:namespace/:kind/:name"
++  element={<CatalogEntityPage />}
++>
++  <EntityPage />
++</Route>
```

Comment:
This one has both element and children - mistake? if not, need to explain what that is about

Reply:
It's explained above. Was a bit brief about it though as I didn't want this to dive too deep into how react router works.

> `FlatRoutes` only considers the first level of `Route` components in its
children, and provides any additional children to the outlet of the route.